File: R/cdf_coverage.R
#' Calculates coverage based upon the CDF
#'
#' @param cdes: A matrix of conditional density estimates; each row
#'   corresponds to an observation, each column corresponds to a grid
#'   point
#' @param z_grid: A vector/matrix of grid points at which the CDE is
#'   estimated
#' @param z_test: The true responses associated with the observations
#' @return A vector of "p-values"
#' @examples
#'   set.seed(12345)
#'   n_grid <- 10001
#'   n_test <- 73
#'   z_grid <- seq(-5, 5, length.out = n_grid)
#'   z_test <- rnorm(n_test)
#'   cdes <- matrix(dnorm(z_grid), nrow = n_test, ncol = n_grid, byrow = TRUE)
#'   pvals <- cdf_coverage(cdes, z_grid, z_test)
#'   ks.test(pvals, "punif")
#' @export
cdf_coverage <- function(cdes, z_grid, z_test) {
  stopifnot(nrow(cdes) == length(z_test))
  stopifnot(ncol(cdes) == length(z_grid))
  z_min <- min(z_grid)
  z_max <- max(z_grid)
  z_delta <- (z_max - z_min) / length(z_grid)

  n_test <- nrow(cdes)
  pvals <- rep(NA, n_test)
  for (ii in seq_len(n_test)) {
    pvals[ii] <- z_delta * sum(cdes[ii, z_grid > z_test[ii]])
  }

  return(pvals)
}

File: python/cdf_coverage.py
import numpy as np


def cdf_coverage(cdes, z_grid, z_test):
    """
    Calculates coverage based upon the CDF

    @param cdes: a numpy array of conditional density estimates;
        each row corresponds to an observation, each column corresponds to a grid
        point
    @param z_grid: a numpy array of the grid points at which cde_estimates is evaluated
    @param z_test: a numpy array of the true z values corresponding to the rows of cde_estimates

    @returns A numpy array of values
    """

    nrow_cde, ncol_cde = cdes.shape
    n_samples = z_test.shape[0]
    n_grid_points = z_grid.shape[0]

    if nrow_cde != n_samples:
        raise ValueError("Number of samples in CDEs should be the same as in z_test."
                         "Currently %s and %s." % (nrow_cde, n_samples))
    if ncol_cde != n_grid_points:
        raise ValueError("Number of grid points in CDEs should be the same as in z_grid."
                         "Currently %s and %s." % (nrow_cde, n_grid_points))

    z_min = np.min(z_grid)
    z_max = np.max(z_grid)
    z_delta = (z_max - z_min)/n_grid_points
    vals = [z_delta * np.sum(cdes[ii, np.where(z_grid > z_test[ii])[0]]) for ii in range(n_samples)]
    return np.array(vals)
